concurrency/conc_and_threading.cpp

```diff
@@ -0,0 +1,142 @@
+#include <iostream>
+#include <filesystem>
+#include <future>
+#include <string>
+#include <vector>
+#include <fstream> 
+
+
+void clientJob(std::exception_ptr& ex_ptr, std::promise<std::filesystem::path> file_name, std::future<std::vector<char>> what_the_server_read);
+void serverJob(std::exception_ptr& ex_ptr, std::future<std::filesystem::path> client_file_name, std::promise<std::vector<char>> promise_file_data);
+
+void readFileName(std::promise<std::filesystem::path>& file_name);
+void readFileContent(std::promise<std::vector<char>>& promise_file_data, std::filesystem::path& file_name);
+
+int main() {
+    
+    std::promise<std::filesystem::path> promise_file_name;
+    std::future<std::filesystem::path> future_file_name = promise_file_name.get_future();
+
+
+    std::promise<std::vector<char>> promise_file_content;
+    std::future<std::vector<char>> future_file_content = promise_file_content.get_future();
+
+    std::exception_ptr client_thread_ex_ptr;
+    std::exception_ptr server_thread_ex_ptr;
+
+    try {
+        std::thread client_thread(clientJob, std::ref(client_thread_ex_ptr), std::move(promise_file_name), std::move(future_file_content));
+        std::thread server_thread(serverJob, std::ref(server_thread_ex_ptr), std::move(future_file_name), std::move(promise_file_content));
+
+        if (client_thread.joinable()) {
+        client_thread.join();
+        }
+        if (server_thread.joinable()) {
+            server_thread.join();
+        }
+    } catch (std::system_error& e) {
+        std::cerr << "failed to create a thread: " << e.what() << std::endl;
+    }
+
+    if (client_thread_ex_ptr != nullptr) {
+        try {
+            std::rethrow_exception(client_thread_ex_ptr);
```

Comment:
why no std::promise?

Reply:
do you mean to not use exception_ptr but to use promise and future for the exceptions in the threads ?

Reply:
yeah I think is much more elegant

Reply:
i preferred the exception_ptr because i think it is a lighter mechanism (doesnt have a lock involved).
in addition i only need to track an exception, while in the promise future mechanism the main idea is to send data and the exception isn't the main purpose. 
let's consider there wasn't an exception to throw in the promise future mechanism i would still need to put a value and not act upon in the main thread.